src/schema.ts

```diff
@@ -625,13 +625,13 @@ export default abstract class BaseSchema<
   }
 
   when(options: ConditionOptions<this>): this;
-  when(keys: string | string[], options: ConditionOptions<this>): this;
+  when(keys: string | readonly string[], options: ConditionOptions<this>): this;
   when(
-    keys: string | string[] | ConditionOptions<this>,
+    keys: string | readonly string[] | ConditionOptions<this>,
     options?: ConditionOptions<this>,
   ) {
     if (!Array.isArray(keys) && typeof keys !== 'string') {
-      options = keys;
+      options = keys as ConditionOptions<this>;
```

Comment:
typescript 4.1.2 is able to infer the correct type of `keys`, given the type guards. Let me know if you prefer to update the typescript version rather than this type assertion.